### ISSUE TITLE
New CS proposal: Cookie Theft Detection Cheat Sheet

### ISSUE THREAD

**Issue by Jxck** [OWASP/CheatSheetSeries#1547]
## What is the proposed Cheat Sheet about?
The way to detect compromised Cookie, for avoid Session Hijacking.

## What security issues are commonly encountered related to this area?
Since authorization process has been updated via Passkey, 2FA etc,
It's so difficult to attack services with compromised Passwords.
So attacker sifting vector to steal Session Cookie and use it to hijack user session.
It saids that Pass the Cookie attack and leaked Cookies are traded at Dark Net.

See more details: [Cookie stealing: the new perimeter bypass – Sophos News](https://news.sophos.com/en-us/2022/08/18/cookie-stealing-the-new-perimeter-bypass/)

It's so difficult to detect Cookie Theft attack since there are no difference between
Real User & Attacker while they sends same valid session cookie.

I believe every services are affected by this kind of attack but there are less informed how to protect.

Recently, Slack shares their own knowledge to prevent this attack.

[Catching Compromised Cookies - Engineering at Slack](https://slack.engineering/catching-compromised-cookies/)

I wonder lot's of big services already doing the prevention like above, but less informed to public.

OWASP cheat sheet is the best place to publish knowledge around here.

## What is the objective of the Cheat Sheet?

No guideline around Cookie Theft attack in current cheat sheet.

## What other resources exist in this area?

Slack shared their experience and it seems good resource to start discussion.

[Catching Compromised Cookies - Engineering at Slack](https://slack.engineering/catching-compromised-cookies/)

**Comment by szh** [comment]
This is really interesting. Are there any other orgs that have discussed if/how they're handling this threat? I'd like to see multiple sources for a more complete view rather than relying on just one company.

**Comment by Jxck** [comment]
@szh Good Question and I can't find.
I believe GAFAM-ish company doing the protection like this but there're no outreach around here (I imagine it's blocked by security reason).
Slack's blog is the first resources I've ever seen, thanks for Slack!

I hope the discussion here will gather knowledge, merged into guideline and share to every services who afraid cookie theft.

**Comment by Jxck** [comment]
I wanna loop in author of Slack's block here @glitched

**Comment by Glitched** [comment]
Hey! Glad to hear you found the blog post useful. I'm not speaking on behalf of Slack here, just my personal interest in the problem:

Protecting against a compromised endpoint is a hard problem, but there are a lot of ways to make progress without as complex a solution as detailed in out blog post. First, I'd suggest a reasonable expiration time on cookies. If a cookie is only valid for a week, the odds are it's still valid when an attacker tries it will be much lower. (From my understanding, the people operating the malware are often not the same people using the cookies, so a time delay may be introduced.) Anything is better than having valid cookies sitting around for months.

Another mitigation might be storing info from the user agent on the session object. If a user created the session on Chrome, I wouldn't expect to see that cookie coming from Edge. I might expect to see a token from an older version of Chrome being used in a newer version of Chrome, but not the other way around. If the user started on a Mac, I wouldn't expect to see the session coming from a Windows machine. It might be worth calling out that iOS devices can sometimes present as macs if you set the "use desktop view" option.

Alternatively, consider teleportation checks. If a user attempts to use a session with an IP address with a location far away from where the session was created, it might be reasonable to ask the user to reauthenticate (or 2FA again).

All of those solutions are definitely incomplete, but they're not nothing! I'm very excited about the [Device Bound Session Credentials (DBSC) proposal](https://github.com/WICG/dbsc?tab=readme-ov-file#introduction) from Google, which aims to prevent sessions from leaving a particular device using hardware TPMs. The Chromium blog has [high-level a overview](https://blog.chromium.org/2024/04/fighting-cookie-theft-using-device.html), too. Microsoft also proposed an addon to DBSC, [Device Bound Session Credentials for Enterprise](https://github.com/WICG/dbsc/blob/main/DBSCE/Overview.md#overview). I'm not sure if the timelines are still on track, but the Github repo says they'd like to have a DBSC trial working by the end of the year.

In the meantime, I'm interested in [Session-Lock](https://session-lock.keyri.com/), which aims to do something similar.

**Comment by Jxck** [comment]
Thanks for your additional information !

Yes, DBSC is the fundamental solution for this kind of attack, but it's just early stage to standardize.  During the work go forward and every device will have TPM, Server side solution should cover problem.

Checking session meta information ((geo)-IP, UA, Sec-Fetch-* etc) could mitigate some attack, even if they can't manage timestamp like Slack does. I think it's can be a 1st Tier recommendation is this Cheat Sheet. And hope some other additional Tier as Defense in Depth manner.

> In the meantime, I'm interested in [Session-Lock](https://session-lock.keyri.com/), which aims to do something similar.
TIL Session-Lock, thanks !

**Comment by Jxck** [comment]
I started drafting roughly.

https://docs.google.com/document/d/1eENyNg9ivICHQpUp6OyLFoarNZXoM8Bfk9__3cTTwXI/edit?tab=t.0

Feel free to comment me.

**Comment by jmanico** [comment]
There are several problems with this work as is.

1) We have cookie guidance in other cheatsheets, I would like this brief work to go there
2) There are comments like "session theft is basically the same as credential theft" which is not true. I may steal your session but without your credentials I cannot do re-authentication challenges.

This needs to be matured in a few ways before we can use it.

**Comment by Jxck** [comment]
@jmanico 

> 1. We have cookie guidance in other cheatsheets, I would like this brief work to go there

I don't against that. (I think you mean [Session Management](https://cheatsheetseries.owasp.org/cheatsheets/Symfony_Cheat_Sheet.html))
I can send PR for session management cheat sheet. but I'm worried that the document might be too long. 

> 2. There are comments like "session theft is basically the same as credential theft" which is not true. I may steal your session but without your credentials I cannot do re-authentication challenges.

Yes, it's meant for "stealing session". I updated with adding "util it's expires".

**Comment by mackowski** [comment]
@jmanico what do you think about @Jxck response to your questions?

**Comment by jmanico** [comment]
Ok let's do a new cheatsheet then. This really is a specialized topic. I agree it might be too long.

**Comment by kwwall** [comment]
> Ok let's do a new cheatsheet then. This really is a specialized topic. I agree it might be too long.

I agree that it may be too long, but rather than placing it in a _separate_ cheat sheet, why not just put the content in a new _collapsible_ section under the [Cookies section of the Session Management Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/Session_Management_Cheat_Sheet.html#cookies)? We'd inevitably want a link to this new content from there anyhow.

## Why my objection?
I am concerned about the proliferation of  new cheat sheets for specialization purposes. More isn't always better (says Mr TL;DR. :smirk:) Before we had collapsible sessions, this made sense, but I think it makes less sense once GitHub started supporting those. (And we still can put an 'cookie_theft' anchor tag there of someone wishes to link directly to the content.) The reason for my concern is that I'm now at my 3rd employer where the company writes their own custom vulnerability templates for various vulnerability categories. And at all 3 companies, the Additional Resources section of the ones I and most other colleagues written almost always had links to these 3 document sources:

1. OWASP Cheat Sheet
2. OWASP Top Ten reference (when applicable)
3. MITRE CWE reference

So, let me just say, I'd rather have a link to only a single OWASP Cheat Sheet than to have one with 2 or 3 links to it as references.  Maybe it's just me being lazy or the rest of you really enjoy having a plethora of references, but I think the developer community is just frustrated by that. We're not writing references for a doctoral dissertation to impress faculty members, so more isn't better, unless your bound and determined to persist the myth that developers can never really understand AppSec because it's too complex and they just aren't smart enough. It that's your take though, then "welcome to the priesthood of the AppSec clerics".  I for one call bullshit on that myth. Let us not forget that we are writing for the primary audience of _developers_. InfoSec / AppSec peeps are just a secondary audience.

## Parting words
I know my mini-rant is not going to convince anyone, but I had to write it just to get it off my chest. You're all gonna do what you would had I not written the this reply. I want you all to know that I have the utmost respect for all of you and I don't think any of you have a giant ego or anything like than I, so it's not because of that.  But I do think we need to collectively step bad and look at maybe do some planned organization of the cheat sheets such as splitting them into tracks or something. I just don't want the Cheat Sheet series to grow so complex that we need a cheet sheet to navigate the cheat sheets.

Just my $03.

**Comment by mackowski** [comment]
@kwwall this is very valuable, I think we can work on some solution to have some 'tracks' or somehow connect the cheatsheets together when they are related.
Before we do that I think we can add a reference/link to this new cheatsheet from old one and vice-versa. And We can add this somwhere in the beginning of the cheatsheet or with a H1 header to make it visible.

**Comment by mackowski** [comment]
@Jxck @jmanico @kwwall @szh do we want to move it forward? 
I think @Jxck can translate docs to markdown and create a PR. Next we need to do the review and add linking between cheatsheets to address that it will be is a separate cheatsheet.
Long term we need to think about 'tracks' or tags in cheatsheets

**Comment by Jxck** [comment]
> do we want to move it forward?

Yes!

> I think @Jxck can translate docs to markdown and create a PR. 

OK, I'll brushing up a bit and create PR later.

@mackowski Is it Okey to send PR as a new Cheatsheet section ?

**Comment by mackowski** [comment]
@Jxck Yes it is okey!

**Comment by kwwall** [comment]
I'm fine if @Jxck does this  PR as a _new_ Cheat Sheet. My mini-rant above wasn't oriented at this specific PR but rather the recent pattern of specialization that I've had recently observed back when I wrote it.

If this new PR is short, we can always fold it into the Session Management CS as a new collapsible section. If it's long, then we can just link to it from the Session Management CS; but I do think we need to do one of those. But we should wait to see how long it is before deciding that.

That said, I think we do need to keep in mind the comments I just wrote in https://github.com/OWASP/CheatSheetSeries/issues/1143#issuecomment-261463

**Comment by Pastekitoo** [comment]
Hi, I would like to work on this one. Can you please assign this to me ?

**Comment by Jxck** [comment]
It's WIP on my side.

Jxck

2025年2月4日(火) 18:15 Pastekitoo ***@***.***>:

**Comment by Pastekitoo** [comment]
@Jxck Do you need some help ?

**Comment by Jxck** [comment]
@Pastekitoo Review's are welcome on my PR!

**Comment by Jxck** [comment]
I'm updating draft on Google Docs before I move this into markdown.
Because I require comment for below

Q. Device Bound Session Credentials are best solution for Cookie Theft Detection here (I believe). But it's been under discussion for now, and It's not standard for now. It can be mentioned in this draft now ? just add a link or explain deeply why it's can be a solution ?

**Comment by Jxck** [comment]
Sent a PR #1615  with explain overview of DBSC.
